site-src/geps/gep-851.md

```diff
@@ -0,0 +1,102 @@
+# GEP-851: Allow Multiple Certificate Refs per Gateway Listener
+
+* Issue: [#851](https://github.com/kubernetes-sigs/gateway-api/issues/851)
+* Status: Implementable
+
+## TLDR
+
+Replace `CertificateRef` field with a `CertificateRefs` field in Gateway
+Listeners.
+
+## Goals
+
+* Allow certs to be attached for multiple unique Hostnames for the same
+  Listener. This is most helpful in the following situations:
+  * The hostname on the Listener is empty or unspecified allowing Routes to
+    attach arbitrary hostnames to the Listener with Routes.
+  * The hostname on the Listener is a wildcard and Route owners want to attach
+    multi-level subdomains. For example, `*.example.com` on the Listener and
+    `app.dev.example.com` on a Route.
+* Enable specifying RSA and ECDSA certs on the same Listener.
+* Improve cert upgrades by allowing new and old certs to be attached to the same
+  Listener.
+
+## Non-Goals
+
+* Be too prescriptive about how multiple certificates are handled for a Gateway
+  Listener. Each implementation may handle this slightly differently.
+
+## Introduction
+
+When we removed certificate references from HTTPRoute ([GEP
+746](/geps/gep-746/)) we lost the capability to attach multiple certificates to
+a Listener. This GEP proposes adding that capability back, directly to the
+Gateway Listener this time.
+
+In addition to helping with the self service approach, where app owners may
+attach multiple hostnames (and certs) to a single Listener, this can help with
+other use cases:
+
+* Attaching RSA and ECDSA certs.
+* Attaching both new and old certs during an upgrade.
+
+## API
+
+The `CertificateRef` field in `GatewayTLSConfig` would be replaced with the
+following `CertificateRefs` field:
+
+```go
+    // CertificateRefs contains a series of references to Kubernetes objects that
+    // contains TLS certificates and private keys. These certificates are used to
+    // establish a TLS handshake for requests that match the hostname of the
+    // associated listener.
+    //
+    // When multiple TLS certificates are specified, SNI matching MUST be used
```

Comment:
So for an Envoy implementation, is the expectation that we would take a listener like:
```
hostname: "*"
certs:
 - foo.com
 - bar.com
 - baz.com
```

is a syntactical sugar for
```
- hostname: foo.com
   certs:
   - foo.com
- hostname: bar.com
   certs:
   - bar.com
```

(although the route binding is slightly different with  `sectionName` match)

Reply:
I think that's correct in that case, but I also think that we'll need to be pretty specific with rules to cover all the possible cases.